tracer/src/Datadog.Trace/AppSec/ApiSecurity.cs

```diff
@@ -6,44 +6,85 @@
 #nullable enable
 using System;
 using System.Collections.Generic;
-using System.IO;
-using System.IO.Compression;
-using System.Text;
-using Datadog.Trace.AppSec.Waf;
-using Datadog.Trace.Sampling;
-using Datadog.Trace.Vendors.Newtonsoft.Json;
-using Datadog.Trace.Vendors.Serilog;
+using System.Collections.Specialized;
+using System.Runtime.CompilerServices;
+using Datadog.Trace.ExtensionMethods;
+using Datadog.Trace.Logging;
 
 namespace Datadog.Trace.AppSec;
 
 internal class ApiSecurity
 {
+    private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<ApiSecurity>();
+    private readonly int _maxRoutesSize;
     private readonly bool _enabled;
-    private readonly OverheadController _overheadController;
+    private readonly TimeSpan _minTimeBetweenReprocessTimeSpan;
+    private readonly OrderedDictionary _processedRoutes = new();
+    private readonly Dictionary<string, bool> _apiSecurityAddress = new() { { "extract-schema", true } };
 
-    public ApiSecurity(SecuritySettings securitySettings)
+    public ApiSecurity(SecuritySettings securitySettings, int maxRouteSize = 4096)
     {
-        _overheadController = new(securitySettings.ApiSecurityMaxConcurrentRequests, (int)(securitySettings.ApiSecuritySampling * 100));
         // todo: later, will be enabled by default, depending on if Security is enabled
         _enabled = securitySettings.ApiSecurityEnabled;
+        _minTimeBetweenReprocessTimeSpan = TimeSpan.FromSeconds(securitySettings.ApiSecuritySampleDelay);
+        _maxRoutesSize = maxRouteSize;
     }
 
-    public bool TryTellWafToAnalyzeSchema(IDictionary<string, object> args)
+    public bool ShouldAnalyzeSchema(bool lastWafCall, Span localRootSpan, IDictionary<string, object> args, string? statusCode, IDictionary<string, object>? routeValues)
     {
-        if (_enabled && _overheadController.AcquireRequest())
+        if (_enabled && lastWafCall && (localRootSpan.Context.TraceContext?.SamplingPriority > (int?)SamplingPriority.AutoReject || localRootSpan.Context.SamplingPriority > (int?)SamplingPriority.AutoReject))
```

Comment:
I saw the RFC for this and it says

> Currently, all tracers should use head-based sampling: sampling priority [...] is decided at trace creation and is then available.

This is not entirely correct. I left a comment in that document, but will copy it here:

Current APM sampling RFC says sampling decision has to be delayed as long as possible, until we need to propagate it (for example, when injecting into headers to a downstream service, or when sending the trace to the agent). This is to take into account information that is not available at trace creation. Some tracers were already delaying the sampling decision even before this last RFC.

Reply:
I left another comment in that thread.

> I see now that this issue may already be covered further down (although for different reasons):
>
> From the RFC:
> > To avoid doing unnecessary computations, tracers must... Delay schema computation until the end of the request... This caters for the sampling decision override, however unlikely